### ISSUE TITLE
NonbondedForce Exceptions Change Energy and Force with Nonbonded Cutoff

### ISSUE THREAD

**Issue by yudongqiu** [pandegroup/openmm#2310]
### Description
In a TIP3P water simulation with `NonbondedForce`, if an exception is added in place of a non-bonded interaction, with the values same as the original parameters, the energy and force still changes. This happens when `CutoffNonPeriodic`, `NonbondedMethod` of `PME` or `CutoffNonPeriodic` is used, but does not happen then `NoCutoff` is used.

### Test files
[test_nonbonded_force_exception.tar.gz](https://github.com/pandegroup/openmm/files/3221569/test_nonbonded_force_exception.tar.gz)

### Test results
```
--- Comparing energy and forces with NoCutoff ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -29409.427612781525
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -29409.427577972412
   - CMMotionRemover: 0.0
Force RMSD: 2.5099529241808254e-05 kJ/mol/nm

--- Comparing energy and forces with CutoffNonPeriodic ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -29506.962612694828
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -29508.085987760183
   - CMMotionRemover: 0.0
Force RMSD: 1.7263203122768331 kJ/mol/nm

--- Comparing energy and forces with CutoffPeriodic ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35298.1263378692
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35294.224625581475
   - CMMotionRemover: 0.0
Force RMSD: 2.5144528434182107 kJ/mol/nm

--- Comparing energy and forces with PME ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35411.07729585373
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35406.73373540354
   - CMMotionRemover: 0.0
Force RMSD: 4.999196693128473 kJ/mol/nm
```

**Comment by peastman** [comment]
This question comes up occasionally.  When using reaction field (which is what you get with CutoffNonPeriodic or CutoffPeriodic), it doesn't apply a cutoff to the nonbonded exceptions.  See #1590 for some discussion of this.  Different applications and force fields make different assumptions about this.  There's no clear "right" answer.  In practice, though, it makes very little difference.

In the case of PME, you "should" get identical forces and energies either way.  The fact that you don't is more of a numerical issue: you're transferring a calculation from reciprocal space to direct space, so you're calculating the same quantity with different approximations.  The difference goes down as you make the error tolerance smaller.  If I add `ewaldErrorTolerance=5e-5` to the call to `createSystem()`, the force and energy differences are much smaller.

```
--- Comparing energy and forces with PME ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35413.33492776551
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35412.64590482105
   - CMMotionRemover: 0.0
Force RMSD: 0.23558059291575073 kJ/mol/nm
```

**Comment by yudongqiu** [comment]
@peastman Thanks a lot for the explanation! I will try `ewaldErrorTolerance=5e-5` with PME and see if it improves my free energy estimation.

**Comment by yudongqiu** [comment]
@peastman Sorry to bother you again, but it seems that adding `ewaldErrorTolerance=5e-5` made the RMSD even a bit higher in my test:
```
--- Comparing energy and forces with PME ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35413.426540550834
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35409.042872482445
   - CMMotionRemover: 0.0
Force RMSD: 5.000542004231953 kJ/mol/nm
```
The new script and testing files are attached:
[test_nonbonded_force_exception_eward.tar.gz](https://github.com/pandegroup/openmm/files/3234719/test_nonbonded_force_exception_eward.tar.gz)

Could you please take a look?

**Comment by peastman** [comment]
I think I see what's going on.  When I tested before I was using my laptop, so I used the CPU platform.  That has good agreement between the two cases when you make the Ewald error tolerance smaller.  But I confirmed that CUDA gives a bigger difference even with a small tolerance.

So I looked to see what's different between the platforms, and I think I've figured it out.  Nonbonded exceptions take some special handling with PME, since there's no way to make the reciprocal space calculation use the altered parameters.  So instead we need three steps:

1. Calculate the reciprocal space energy, which ignores exceptions.
2. In direct space, subtract off the energy that was incorrectly included in step 1 for each exception.
3. Again in direct space, add in the full energy for each exception.

Step 2 is handled differently in the two platforms.  The CUDA platform does it as part of the main direct space nonbonded kernel, whereas the CPU platform does it separately as a loop over exceptions.  This leads to the following difference: the CUDA platform only subtracts off the reciprocal contribution for excluded interactions if they are *closer than the cutoff distance*.

Normally this isn't an issue.  The main use of exceptions is for 1-4 interactions, which are well within the cutoff since the atoms are connected by bonds.  But your case is different, because you're adding exceptions between one water and every other water in the system, even ones that are far away.

To confirm this hypothesis, I modified the CPU platform to only subtract the reciprocal contribution for atom pairs within the cutoff.  Sure enough, that led to a force discrepancy similar to the one for CUDA.

I'm looking into how I can fix this in the next release.  In the mean time there might be a workaround that will help with your particular use case.  Can you describe exactly what you're using the exceptions for?

**Comment by yudongqiu** [comment]
@peastman Thanks a lot for the detailed explanation!
I'm writing a script to estimate the solvation free energy. As part of the system setup, I would like to separate the solute-solvent interactions from the others, therefore I tried to use exceptions in the NonbondedForce.

As part of the system setup, I tried to use the ExceptionParameterOffsets to generate the alchemical intermediate states, with zero LJ and a reduced coulomb interaction. However I found this approach lead to an over-estimate of the free energy.

In another approach, I was able to get a good free energy value, by setting both the LJ and coulomb parameters to zero in NonbondedForce to complete exclude them, then use another CustomNonbondedForce for building the alchemical intermediate states. The only concern is that CustomNonbondedForce does not support PME, so the reduced coulomb interactions may not be as accurate as the first approach.

**Comment by leeping** [comment]
Hi Peter,

Yudong in my group wrote this script to compute solvation free energies by decoupling the solute-solvent nonbonded interactions.  In our first approach (call this Approach 1), we wanted to use PME for all electrostatics and decouple the solute-solvent pairwise Coulomb interactions using exceptions.  I thought this would be the preferred approach because it corresponds most closely to 'non-free-energy' simulations.

As Yudong mentions, we have an alternative solution (Approach 2) that uses CustomNonbondedForce with a cutoff for the solute-solvent pairwise Coulomb interactions.  We found that gives good results consistent with the literature, though I was slightly uncomfortable that we were using "short-range" electrostatics for the solute-solvent interactions.  

Now that you've explained how OpenMM handles pairwise exceptions in PME, I am questioning if Approach 1 is a good route or not.  Even if you make it possible to add "exception corrections" for distant pairs in direct space, we would still need to do pairwise exceptions for the periodic interactions.  I'm not familiar enough with PME to know whether this is possible theoretically.  We'd like to avoid a situation where the "fully decoupled" simulation still has full interactions between the solute and the non-neighboring periodic images of the solvent.

**Comment by peastman** [comment]
Yes, that's an important point.  Exceptions are designed to be used between atoms that are bonded to each other (e.g. 1-4 interactions).  They are applied only between the two specific atoms that are bonded, not between other periodic copies of those atoms.  When you alter the parameters for a 1-4 interaction, that doesn't affect how each atom interacts with periodic copies of the other atom, which after all are in different molecules.

**Comment by leeping** [comment]
Hi Peter,

I think a better solution to my problem would be to append an extra state to my list of alchemical intermediates corresponding to the fully interacting state that uses PME for all Coulomb interactions. Using the notation (lambda_coulomb, lambda_LJ), we would then have:

(1.0*, 1.0*)
(1.0, 1.0)
(0.9, 1.0)
...
(0.0, 1.0)
(0.0, 0.9)
...
(0.0, 0.0)

The (1.0*, 1.0*) state with the asterisks has the fully interacting solute and solvent using PME for all Coulomb interactions, i.e. the one that we cannot reproduce by using exceptions. 

The (1.0, 1.0) state would then have fully interacting solute and solvent using CustomNonbondedForce for solute-solvent Coulomb interactions, which is not exactly the endpoint we want, but should have good thermodynamic overlap with the (1.0*, 1.0*) state.

Within our scheme, the (0.0, 0.0) state would have the solute and solvent fully decoupled, but the solute-solute and solvent-solvent Coulomb interactions are still using PME because we created two instances of NonbondedForce.

I think this would solve our problem without requiring any changes to OpenMM. What do you think?

**Comment by peastman** [comment]
That sounds plausible.  Give it a try and see how it works.

**Comment by peastman** [comment]
The fix is in #2318.  Could you verify that it fixes the problem for you?

**Comment by yudongqiu** [comment]
@peastman Thanks a lot for the fix! I just checked that it did fix the problem with PME:
```
--- Comparing energy and forces with PME ---
Original System
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35413.26848228535
   - CMMotionRemover: 0.0
New System with Exceptions
 * Energy Decomposition in kJ/mol
   - HarmonicBondForce: 0.0
   - HarmonicAngleForce: 0.0
   - NonbondedForce: -35412.58433533693
   - CMMotionRemover: 0.0
Force RMSD: 0.23559729529354673 kJ/mol/nm
```
